js/app/modules/selection/xapian.js

```diff
@@ -85,24 +85,43 @@ const Xapian = new Module.Class({
                 return;
             }
 
-            let next_query = info.more_results;
-            if (!next_query) {
+            let results_added = results.filter(this.add_model, this);
+            let need_more_results = num_desired > results_added.length;
+
+            let more_results = info.more_results;
+            if (!!more_results) {
+                // If we got `more_results`, check to see if we're at the limit
+                let can_load_more = info.upper_bound > results.length;
+                if (can_load_more !== this._can_load_more) {
+                    this._can_load_more = can_load_more;
+                    this.notify('can-load-more');
+                }
+
+                // Only recurse if we need more results and can get more
+                if (need_more_results && this._can_load_more) {
+                    this._next_query = more_results
+                    this.queue_load_more(num_desired - results_added.length);
+                }
+            } else {
+                // No `more_results` - see if we have additional queries to fill
+                // the results
                 this._query_index++;
-                next_query = this.construct_query_object(num_desired, this._query_index);
-            }
+                this._next_query = this.construct_query_object(num_desired,
+                                                               this._query_index);
+
+                // We have a new query. We only assume that we can get more results
+                // if we filled up the desired count. If the next query isn't defined,
+                // we know for sure that we won't be able to fill the results array.
+                let can_load_more = !need_more_results || !!this._next_query;
```

Comment:
It somewhat of a tricky logic here and I'm not sure if the comment applies but I'll be happy to change it if needed.

In essence, if the query didn't get us all the results we wanted and we don't have a new query, there should be no case where re-running that same query will give us more new results so `can_load_more` in those cases should be `false`. If we don't do this check, if we got partial results we will have "can-load-more" buttons enabled when there's no more data to fetch which is imo an error.